usefulscripts/dump_nca.js

```diff
@@ -0,0 +1,349 @@
+sc.getFSPPR = function () {
+	if (sc.closed_pr !== undefined) {
+		return;
+	}
+	sc.enableTurbo();
+	var i = 0;
+	var srv = null;
+	while (true) {
+		sc.ipcMsg(2).setType(5).sendTo('pm:shell');
+		var srvResult = sc.getService("fsp-pr");
+		if(srvResult.isOk) {
+			srv = srvResult.getValue();
+			break;
+		}
+		i++;
+	}
+	utils.log('Got fsp-pr handle after ' + i + ' iterations: ');
+	utils.log('fsp-pr handle: 0x' + srv.toString(16));
+	sc.svcCloseHandle(srv).assertOk();
+	sc.closed_pr = true;
+};
+
+sc.getFSPPR();
+sc.enableTurbo();
+
+if(sc.pr_handle) {
+	sc.svcCloseHandle(sc.pr_handle);
+	sc.pr_handle = undefined;
+}
+
+sc.getService("fsp-pr", (fsppr) => {
+	var pid = sc.getService('fsp-srv', (tmp_hnd) => {
+		utils.log("got fspsrv");
+		sc.ipcMsg(1).sendPid().data(0).sendTo(tmp_hnd).assertOk();
+		return sc.read4(sc.ipcBufAddr, 0xC >> 2);
+	});
+	utils.log('Got process PID: '+pid.toString(16));
+	
+	var buf1_sz = 0x1C;
+	var buf2_sz = 0x2C;
+	var buf = sc.malloc(buf1_sz + buf2_sz);
+	var buf2 = utils.add2(buf, buf1_sz);
+	
+	// buffer init
+	sc.write4(1, buf, 0x0>>2);
+	sc.write8([0xFFFFFFFF, 0xFFFFFFFF], buf, 0x4 >> 2); // This is the permissions value.
+	sc.write4(buf1_sz, buf, 0xC >> 2);
+	sc.write4(buf1_sz, buf, 0x14 >> 2);
+	
+	sc.write4(1, buf2, 0x0 >> 2);
+	sc.write8([0xFFFFFFFF, 0xFFFFFFFF], buf2, 0x4 >> 2); // This is the permissions value -- actual perms = buf2_val & buf1_val
+	sc.write4(0xFFFFFFFF, buf2, 0x14 >> 2);
+	sc.write4(0xFFFFFFFF, buf2, 0x18 >> 2);
+	sc.write4(0xFFFFFFFF, buf2, 0x24 >> 2);
+	sc.write4(0xFFFFFFFF, buf2, 0x28 >> 2);
+	
+    /* Change to mount a particular title's romfs */
+	var tid = '0000000000000000';
+	
+	sc.ipcMsg(256).data(0).sendTo(fsppr).assertOk().show();
+	sc.ipcMsg(1).data(pid).sendTo(fsppr).assertOk().show();
+	sc.ipcMsg(0).data(2, [pid,0], utils.parseAddr(tid), buf1_sz, buf2_sz, pid, pid, 0, 0, 0, 0, 0).aDescriptor(buf, buf1_sz).aDescriptor(buf2, buf2_sz).sendTo(fsppr).assertOk().show();
+	sc.free(buf);
+	sc.free(buf2);

```

Comment:
This is a bug.  buf2 was not separately `malloc()`'d, but is part of buf.  therefore, line 64 needs to be removed.